std/json.zig

```diff
@@ -0,0 +1,1308 @@
+
+    const object_bit     = 0;
+    const array_bit      = 1;
+    const max_stack_size = @maxValue(u8);
+
+    pub fn init() StreamingJsonParser {
+        var p: StreamingJsonParser = undefined;
+        p.reset();
+        return p;
+    }
+
+    pub fn reset(p: &StreamingJsonParser) void {
+        p.state = State.TopLevelBegin;
+        p.count = 0;
+        // Set before ever read in main transition function
+        p.after_string_state = undefined;
+        p.after_value_state = State.ValueEnd;   // handle end of values normally
+        p.stack = 0;
+        p.stack_used = 0;
+        p.complete = false;
+        p.string_has_unicode_escape = false;
+        p.number_is_integer = true;
+    }
+
+    pub const State = enum {
+        // NOTE: @tagName determines the name based on index, not value.
+
+        // These must be first with these explicit values as we rely on them for indexing the
+        // bit-stack directly and avoiding a branch.
+        ObjectSeparator = 0,
+        ValueEnd = 1,
+
+        TopLevelBegin,
+        TopLevelEnd,
+
+        ValueBegin,
+        ValueBeginNoClosing,
+
+        String,
+        StringUtf8Byte3,
+        StringUtf8Byte2,
+        StringUtf8Byte1,
+        StringEscapeCharacter,
+        StringEscapeHexUnicode4,
+        StringEscapeHexUnicode3,
+        StringEscapeHexUnicode2,
+        StringEscapeHexUnicode1,
+
+        Number,
+        NumberMaybeDotOrExponent,
+        NumberMaybeDigitOrDotOrExponent,
+        NumberFractionalRequired,
+        NumberFractional,
+        NumberMaybeExponent,
+        NumberExponent,
+        NumberExponentDigitsRequired,
+        NumberExponentDigits,
+
+        TrueLiteral1,
+        TrueLiteral2,
+        TrueLiteral3,
+
+        FalseLiteral1,
+        FalseLiteral2,
+        FalseLiteral3,
+        FalseLiteral4,
+
+        NullLiteral1,
+        NullLiteral2,
+        NullLiteral3,
+
+        // Only call this function to generate array/object final state.
+        pub fn fromInt(x: var) State {
+            std.debug.assert(x == 0 or x == 1);
+            return State(u6(x));
+        }
+    };
+
+    pub const Error = error {
+        InvalidTopLevel,
+        TooManyNestedItems,
+        TooManyClosingItems,
+        InvalidValueBegin,
+        InvalidValueEnd,
+        UnbalancedBrackets,
+        UnbalancedBraces,
+        UnexpectedClosingBracket,
+        UnexpectedClosingBrace,
+        InvalidNumber,
+        InvalidSeparator,
+        InvalidLiteral,
+        InvalidEscapeCharacter,
+        InvalidUnicodeHexSymbol,
+        InvalidUtf8Byte,
+        InvalidTopLevelTrailing,
+        InvalidControlCharacter,
+    };
+
+    // Give another byte to the parser and obtain any new tokens. This may (rarely) return two
+    // tokens. token2 is always null if token1 is null.
+    //
+    // There is currently no error recovery on a bad stream.
+    pub fn feed(p: &StreamingJsonParser, c: u8, token1: &?Token, token2: &?Token) Error!void {
```

Comment:
May be more idiomatic/nicer to return a double-token struct here.

Reply:
I wonder if we could make it like the zig tokenizer, and make it always return at least 1 token? Here's the signature:

```zig
    pub fn next(self: &Tokenizer) Token {
````

Here's what's nice about it: 

 * It can't fail. But it can return Token.Invalid. The invalid token has start and end of the invalid character(s), and then normal tokens can follow.
 * It always returns 1 token.
   - If you get to the end it will always return Token.Eof.
   - If it needs to emit 2 tokens, it saves the one token and feeds it to the caller for next time, making the API simpler for the caller.

Reply:
I just realized, the difference between the zig tokenizer and this json one is that the json one is streaming. So that makes sense.

If the caller already has to handle possibly greater than 1 token though, then would it make more sense for the `feed` function to accept a `buf: []const u8` instead of `c: u8`, and then return N tokens?

...oh then it would need to be a callback, because if we needed to return > N tokens, and the return token slice was too small, we would lose data.

Alright, you're 2 steps ahead of me :-)

Reply:
You could still change this to accept a slice. You instead could take a slice of tokens for output and fill until the input slice runs out or you run out of usable token space. If you run out of token space, return an `error.NeedMoreTokens` and save the current position in the buffer until the caller can use filled tokens and call again.

More complicated with the only real benefit being a saved function call cost. Technically you could perform faster matching of quoted-strings, ends of numbers etc by skipping some state changes but that would be special-casing longer buffer sizes and may have the reverse effect in other cases.